### PR TITLE
API for series data and grids

### PR DESCRIPTION
**Pull request by stefwalter**
Note that this commit isn't really useful by itself. The cockpit.series()
is essentially a mix in and needs another object to drive it.

In addition note that the word "series" is both singular and plural.

### CODE REVIEW

**Comment by stefwalter** [comment]
Found a bug.

**Comment by stefwalter** [comment]
@mvollmer This is ready for review. In addition to adding the various API's I've migrated the CpuMonitor and MemoryMonitor graphs to the new metric channels, grids, and series data.

File: doc/guide/api-cockpit.xml [hunk 1]
```diff
@@ -1862,6 +1862,366 @@ handle = file.watch(callback);
...
+      <para>Creates a grid object to contain series data.</para>
+
+      <para>The <code>interval</code> is the granularity of the grid. Usually this is
+        a number of milliseconds, when used with time series data. The <code>beg</code>
+        and <code>end</code> are the bounds of the grid. If omitted they will be set to
+        zero for an initially empty grid.</para>
+
+      <para>If <code>beg</code>
+        and/or <code>end</code> are negative (including negative zero) then they will be set to the
+        number of intervals prior to the current time taken as an interval.</para>
```
Comment (mvollmer) [inline doc/guide/api-cockpit.xml]: I find this hard to understand.  An example might help.

File: doc/guide/api-cockpit.xml [hunk 2]
```diff
@@ -1862,6 +1862,366 @@ handle = file.watch(callback);
...
+  grid.move(beg[, end])
+</programlisting>
+
+      <para>Move the grid to new <code>beg</code> and <code>end</code> range. Data will be
+        discarded from the rows and <link linkend="cockpit-grid-sync"><code>grid.sync()</code></link>
+        will be called to load or reload series data for the new range of offsets.</para>
+
+      <para>If <code>end</code> is not specified it will be set to <code>beg</code>. If <code>beg</code>
+        and/or <code>end</code> are negative (including negative zero) then they will be set to the
+        number of intervals prior to the current time taken as an interval.</para>
```
Comment (mvollmer) [inline doc/guide/api-cockpit.xml]: Hard to understand.

File: doc/guide/api-cockpit.xml [hunk 3]
```diff
@@ -1862,6 +1862,366 @@ handle = file.watch(callback);
...
+
+    </refsection>
+
+    <refsection id="cockpit-grid-onnotify">
+      <title>grid.onnotify</title>
+<programlisting>
+  $(grid).on("notify", function(index, count) { ... });
+</programlisting>
+
+      <para>An event that is triggered when same part of the series data in grid changes. The
```
Comment (mvollmer) [inline doc/guide/api-cockpit.xml]: s/same/some/

File: doc/guide/api-cockpit.xml [hunk 4]
```diff
@@ -1862,6 +1862,366 @@ handle = file.watch(callback);
...
+            will be retrieved from <ulink url="http://pcp.io">PCP</ulink>if it is
+            available.</para></listitem>
+        </varlistentry>
+      </variablelist>
+    </refsection>
+
+    <refsection id="cockpit-metrics-fetch">
+      <title>metrics.fetch()</title>
+<programlisting>
+  metrics.follow(beg, end)
```
Comment (mvollmer) [inline doc/guide/api-cockpit.xml]: s/follow/fetch/

File: doc/guide/api-cockpit.xml [hunk 5]
```diff
@@ -1862,6 +1862,366 @@ handle = file.watch(callback);
...
+</programlisting>
+
+      <para>Retrieve archived metrics data between <code>beg</code> and <code>end</code>. The
+        arguments can either be numbers, in which case they are interval based offsets, or they
+        can be javascript Date objects.</para>
+    </refsection>
```
Comment (mvollmer) [inline doc/guide/api-cockpit.xml]: This left me with the following question: Why would I call 'fetch'?  Isn't grid.move() the primary mechanisms to trigger data retrieval?

Comment (stefwalter) [inline doc/guide/api-cockpit.xml]: With some series sources you don't even know what rows to add to your grid until you retrieve something. For example with cAdvisor, you have to retrieve data to know which containers are available, and add rows for those containers to your grid, after which grid.move() will retrieve further data for them.

So I wanted there to be a general consistent fetch function available.

File: pkg/shell/cockpit-plot.js [hunk 6]
```diff
@@ -24,7 +24,128 @@ define([
...
+        points: { radius: 0 },
+        grid: { borderWidth: 1,
+                aboveData: true,
+                color: "black",
+                borderColor: $.color.parse("black").scale('a', 0.22).toString(),
+                labelMargin: 0
+              }
+    };
+
+    var num_series = grid.rows.length;
```
Comment (mvollmer) [inline pkg/shell/cockpit-plot.js]: This is unused and a bit confusing.  grid.rows.length is 8, for example, while data.length is only 4.

File: doc/guide/api-cockpit.xml [hunk 7]
```diff
@@ -1862,6 +1862,366 @@ handle = file.watch(callback);
...
+      <title>grid.close()</title>
+<programlisting>
+  grid.close()
+</programlisting>
+
+      <para>Close the grid, and stop updating the rows.</para>
+    </refsection>
+
+    <refsection id="cockpit-grid-rows">
+      <title>grid.rows</title>
```
Comment (mvollmer) [inline doc/guide/api-cockpit.xml]: I think we might be better off not exposing grid.rows as part of the API.  Clients can easily build their own collections of rows, and that is actually cleaner, I think.  See my comment in the cpu monitor code.

File: pkg/shell/cockpit-cpu-status.js [hunk 8]
```diff
@@ -52,28 +48,50 @@ PageCpuStatus.prototype = {
...
+            metrics: metrics,
+            cache: "cpu-status-rate"
+        });
+
+        /* The grid shows us the last five minutes */
+        self.grid = cockpit.grid(1000, -300, -0);
+        metrics.forEach(function(metric) {
+            self.grid.add(self.channel, [ metric.name ]);
+        });
+
```
Comment (mvollmer) [inline pkg/shell/cockpit-cpu-status.js]: From here on, we rely on the fact that grid.add has added the rows in order to grid.rows.  Creating a explicit array is cleaner, I think:

```
var data = metrics.map(function (metric) { return self.grid.add(self.channel, [ metric.name ]); });
```

and then passing `data` to setup_complicated_plot.

File: doc/guide/api-cockpit.xml (hunk 7 above)
Comment (stefwalter) [inline doc/guide/api-cockpit.xml]: I don't mind doing that for now. When a real use case presents itself, we can see how to make it work better.

File: doc/guide/api-cockpit.xml (hunk 1 above)
Comment (stefwalter) [inline doc/guide/api-cockpit.xml]: Very good. Changed.

File: doc/guide/api-cockpit.xml (hunk 2 above)
Comment (stefwalter) [inline doc/guide/api-cockpit.xml]: Replaced with above text.

File: doc/guide/api-cockpit.xml (hunk 3 above)
Comment (stefwalter) [inline doc/guide/api-cockpit.xml]: Changed.

File: doc/guide/api-cockpit.xml (hunk 4 above)
Comment (stefwalter) [inline doc/guide/api-cockpit.xml]: Fixed.

File: pkg/shell/cockpit-plot.js (hunk 6 above)
Comment (stefwalter) [inline pkg/shell/cockpit-plot.js]: Yes, removed.

File: doc/guide/api-cockpit.xml (hunk 7 above)
Comment (stefwalter) [inline doc/guide/api-cockpit.xml]: Done. Also removed a bunch of tests related to this.

File: pkg/shell/cockpit-cpu-status.js (hunk 8 above)
Comment (stefwalter) [inline pkg/shell/cockpit-cpu-status.js]: Changed this around so we pass the rows in the series.

Comment (stefwalter) [inline pkg/shell/cockpit-cpu-status.js]: Making removal of this a separate commit intentionally.

File: doc/guide/api-cockpit.xml (hunk 5 above)
Comment (mvollmer) [inline doc/guide/api-cockpit.xml]: Makes sense.

**Pull request closed without merging**